### PR TITLE
Remove order by from mutations interpreter and add check

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Remove order by stage from mutations because we read from a single ordered part in a single thread. Also add check that the order of rows in mutation is ordered in sorting key order and this order is not violated.

### CODE REVIEW

File: dbms/src/DataStreams/CheckSortedBlockInputStream.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+{
+    children.push_back(input_);
+}
+
+SortDescriptionsWithPositions
+CheckSortedBlockInputStream::addPositionsToSortDescriptions(const SortDescription & sort_description)
+{
+    SortDescriptionsWithPositions result;
+
+    for (SortColumnDescription description_copy : sort_description)
```
Comment (KochetovNicolai) [inline dbms/src/DataStreams/CheckSortedBlockInputStream.cpp]: Possible extra `column_name` copy. (not very important)

File: dbms/src/DataStreams/CheckSortedBlockInputStream.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    : header(input_->getHeader())
+    , sort_description_map(addPositionsToSortDescriptions(sort_description_))
+{
+    children.push_back(input_);
+}
+
+SortDescriptionsWithPositions
+CheckSortedBlockInputStream::addPositionsToSortDescriptions(const SortDescription & sort_description)
+{
+    SortDescriptionsWithPositions result;
```
Comment (KochetovNicolai) [inline dbms/src/DataStreams/CheckSortedBlockInputStream.cpp]: reserve

File: dbms/src/DataStreams/CheckSortedBlockInputStream.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            description_copy.column_number = header.getPositionByName(description_copy.column_name);
+
+        result.push_back(description_copy);
+    }
+
+    return result;
+}
+
+/// Compares values in columns. Columns must have equal types.
+struct SortingLessComparator
```
Comment (KochetovNicolai) [inline dbms/src/DataStreams/CheckSortedBlockInputStream.cpp]: Technically, it's `LessOrEqual` comparator :)

File: dbms/src/DataStreams/CheckSortedBlockInputStream.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+Block CheckSortedBlockInputStream::readImpl()
+{
+    Block block = children.back()->read();
+    if (!block)
+        return block;
+
+    SortingLessComparator less(sort_description_map);
+
+    auto block_columns = block.getColumns();
+    if (!last_row.empty() && !less(last_row, 0, block_columns, 0))
```
Comment (KochetovNicolai) [inline dbms/src/DataStreams/CheckSortedBlockInputStream.cpp]: Funny thing: even if you have checked that block is not empty via `!block`, it doesn't mean that it has any row.
The situation when block has empty columns is valid (but very rare indeed).

File: dbms/src/DataStreams/CheckSortedBlockInputStream.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        throw Exception("Sort order of blocks violated", ErrorCodes::LOGICAL_ERROR);
+
+    size_t rows = block.rows();
+    for (size_t i = 1; i < rows; ++i)
+        if (!less(block_columns, i - 1, block_columns, i))
```
Comment (KochetovNicolai) [inline dbms/src/DataStreams/CheckSortedBlockInputStream.cpp]: Maybe we need de-virtualized version of compare here. But ok for now.

File: dbms/src/DataStreams/CheckSortedBlockInputStream.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            throw Exception("Sort order of blocks violated", ErrorCodes::LOGICAL_ERROR);
+
+    last_row.clear();
```
Comment (KochetovNicolai) [inline dbms/src/DataStreams/CheckSortedBlockInputStream.cpp]: Another way to implement it is to cache the last block.
Then we wouldn't need to copy row every time.
But I think it is not very long to copy single row per block.

**Review by KochetovNicolai — CHANGES_REQUESTED** [review]

**Review by KochetovNicolai — APPROVED** [review]

**Comment by alesapin** [comment]
I think we can merge without all tests, because last fix is just refactoring.

**Pull request merged**